Expensify/App
# IOU Cancel/Decline

CC @iwiznia, @marcaaron, @tgolen, @shawnborton -- no pressure to review, but if you would like to then please try to provide initial feedback this week, as this will help us to remain on track for releasing the newsletter next week 🙇 🎉 

### Details

Following up on the [IOUDetails Modal ](https://github.com/Expensify/Expensify/issues/158811)issue, we need to implement transaction rejection. This PR does does the following:
- Adds transaction rejection logic (labelled as 'Decline' / 'Cancel')
- Hides the Preview Component if the iouReport has all transactions rejected
- Removes the IOUReport's transactions property -- as of this PR I can confirm that they are not necessary as we use reportActions within IOUDetails Modal

https://user-images.githubusercontent.com/10736861/119871798-b7e19000-bf1a-11eb-933a-4787bd5b5881.mov

### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/163797

### Tests

**Internal Testing Notes**
- _Please ensure you update and rebuild Auth before testing, specifically the changes from this [Web-E PR](https://github.com/Expensify/Web-Expensify/pull/31059) and this [Auth PR](https://github.com/Expensify/Auth/pull/5615) are required_
- _Dev can be slow to complete requests! Be aware that you may see 60 seconds requests and failed requests while testing. Restarting the VM and using test accounts with few chatReports helps_
- _We plan to polish the feature with an ActivityIndicator when the cancel/decline buttons are pressed, but this is not required for release_

**Test Steps**
_The screencast above shows the exact test flow_
1. Login with two accounts: `userA` & `userB`, ideally both viewable at once
2. As `userA`, start a new chat with `userB`
3. Request money of any amount from `userB`, via the 'Request Money' flow
4. Make a 2nd request as `userA` of any amount
5. Make a 3rd request as `userB`, ensure that the amount is less than the current total
6. Navigate to the IOUDetails as both users

<img width="1789" alt="Screenshot 2021-05-28 at 14 38 49" src="https://user-images.githubusercontent.com/10736861/119992432-72789d80-bfc2-11eb-99ef-d140153ac29d.png">

7. As `userA`, you should see 2 'Cancel' and one 'Decline' buttons
8. As `userB`, you should see 2 'Decline' and one 'Cancel' button 
9. As `userA`, cancel one of the transactions

<img width="901" alt="Screenshot 2021-05-28 at 14 43 45" src="https://user-images.githubusercontent.com/10736861/119993113-1febb100-bfc3-11eb-9fed-bb9a708fe567.png">

10. The UI should update. A 'Canceled the £X.YZ request' message should appear and the Cancel button should no longer be displayed for that transaction
11. As `userB`, decline one of the transactions
12. The UI should update. A 'Declined the £X.YZ request' message should appear' and the Decline button should no longer be displayed
13. As either user, cancel/decline the 3rd transaction
14. The Preview Comp should no longer be displayed on the IOUDetails Modal OR in the Chat

<img width="1791" alt="Screenshot 2021-05-28 at 14 40 47" src="https://user-images.githubusercontent.com/10736861/119993222-3eea4300-bfc3-11eb-9a60-6f4955f72d9b.png">

15. As `userA`, create a 4th IOU request via 'Request Money'
16. Open the IOUDetailsModal, you should see the previously canceled transactions
17. As `userB`, pay the IOU via any settlement method (you will probably only see the default option)

<img width="1790" alt="Screenshot 2021-05-28 at 14 41 40" src="https://user-images.githubusercontent.com/10736861/119993477-7c4ed080-bfc3-11eb-8cf6-0690dc5d93b8.png">

18. The Preview should remain visible in the IOUDetails Modal and all of the canceled/declined transactions should be visible.
19. Now repeat the steps with other platforms side0by-side (for example: Desktop & iOS, followed by iOS mobile web & Android)

### QA Steps

_Run above test steps_

### Tested On


### Screenshots


#### Web
<img width="1784" alt="Screenshot 2021-05-27 at 18 47 25" src="https://user-images.githubusercontent.com/10736861/119873035-02174100-bf1c-11eb-81be-673bf392614e.png">


#### Mobile Web
![Simulator Screen Shot - iPhone 11 - 2021-05-27 at 18 59 42](https://user-images.githubusercontent.com/10736861/119874813-eca31680-bf1d-11eb-8c1d-bfa9fe4cef1a.png)

#### Desktop
<img width="1201" alt="Screenshot 2021-05-27 at 18 51 29" src="https://user-images.githubusercontent.com/10736861/119873556-91245900-bf1c-11eb-91bd-f1372e7c7b03.png">


#### iOS
![Simulator Screen Shot - iPhone 11 - 2021-05-27 at 18 57 09](https://user-images.githubusercontent.com/10736861/119874311-5e2e9500-bf1d-11eb-9e47-8db39d2baa7f.png)


#### Android
![Screenshot_1622138512](https://user-images.githubusercontent.com/10736861/119874895-08a6b800-bf1e-11eb-8d95-07656b8c7e37.png)
